[CALCITE-1466] Support for UNSIGNED types of TINYINT, SMALLINT, INT, …

mihaibudiu opened the pull request:
…BIGINT in type system

The implementation uses the JOOQ classes for unsigned numbers. These have some shortcomings, but it's better than starting from scratch.

There are many possible choices for the semantics of operations between signed and unsigned, I have chosen one which is similar to MySQL (int + uint = uint). I don't think it's the most intuitive semantics, but I think any choice will be weird in some corner cases.

9 years after the issue has been filed, here is an implementation.
The test coverage is not very high, but there really are many tens of new functions, so in the first commit I tested only a representative sample of these, under the assumption that if one works, they all will.

mihaibudiu commented:
https://issues.apache.org/jira/browse/CALCITE-1466

mihaibudiu commented on core/src/test/resources/sql/unsigned.iq:
```diff
@@ -0,0 +1,126 @@
+!use post
+!set outputformat csv
+
+SELECT CAST(200 AS INT UNSIGNED) - 100;
```
In MySQL you write `UNSIGNED` instead of `INT UNSIGNED`, and then the `INT` tests work.
I can change the parser to accept this variant as well.

NobiGo replied:
Does MySQL only allow conversions of the same type, such as: INT SIGNED -> INT UNSIGNED and INT UNSIGNED -> INT SIGNED? INT SIGNED-> BIGINT UNSIGNED will failed.

strongduanmu commented on core/src/main/codegen/templates/Parser.jj:
```diff
@@ -5903,13 +5904,21 @@ SqlTypeNameSpec SqlTypeName1(Span s) :
         <BOOLEAN> { s.add(this); sqlTypeName = SqlTypeName.BOOLEAN; }
     |
-        ( <INTEGER> | <INT> ) { s.add(this); sqlTypeName = SqlTypeName.INTEGER; }
+        ( <INTEGER> | <INT> ) ( <UNSIGNED> { unsigned = true; })? {
```
Considering that not all databases support `UNSIGNED` types, can we add `conformance.allowUnsignedType()` to perform checks?

strongduanmu commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeAssignmentRule.java:
```diff
@@ -70,33 +70,73 @@ private SqlTypeAssignmentRule(
+    // USMALLINT is assignable from...
+    rule.clear();
+    rule.add(SqlTypeName.UTINYINT);
+    rule.add(SqlTypeName.TINYINT);
```
In MySQL, the range of TINYINT is -128 ~ 127, while the range of USMALLINT is 0 ~ 65535. Considering that TINYINT includes negative numbers, perhaps we cannot assign TINYINT to USMALLINT type.

NobiGo commented on core/src/test/resources/sql/unsigned.iq:
```diff
@@ -0,0 +1,126 @@
+1000
+!ok
+
+SELECT CAST(CAST(10 AS TINYINT) AS INT UNSIGNED);
```
I change this SQL to `SELECT CAST(CAST(1 AS TINYINT) AS  UNSIGNED);` run in MySQL. It throws exception `You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'TINYINT) AS  UNSIGNED)`

mihaibudiu replied:
mysql is very inconsistent, it does not allow you to use tinyint in casts. that's stupid.
nothing to do with this PR.

NobiGo replied:
I noticed you have updated the reference.md. I think it would be sufficient to describe the capabilities we support in the Jira summary.

mihaibudiu commented:
does anyone know why the gradle wrapper validation fails? I could not see anything strange in the logs

suibianwanwank commented:
I've encountered this issue a few times. it seems related to rapid consecutive commits. I think we can ignore it.

xiedeyantu commented:
> > does anyone know why the gradle wrapper validation fails? I could not see anything strange in the logs
> 
> Check the CI info and rerun the failed jobs.

@NobiGo Do we have some commands to rerun the specific ci case?

strongduanmu reviewed (approved):
LGTM +1

mihaibudiu commented:
I plan to merge this PR if there are no more comments.
If we find bugs and missing functionality related to unsigned numbers, we will file additional issues.

mihaibudiu commented:
I had to push another commit, since I discovered attempting to use the unsigned class that sometimes RexBuilder attempted to build unsigned literals, which do not exist.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rex/RexBuilder.java:
```diff
@@ -1943,11 +1959,11 @@ public RexNode makeZeroRexNode(RelDataType type) {
   }
 
-  private static Comparable zeroValue(RelDataType type) {
+  private static Comparable<?> zeroValue(RelDataType type) {
```
at least the IDE is happier if you do.
Comparable is a generic type.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rex/RexBuilder.java:
```diff
@@ -1961,6 +1977,10 @@ private static Comparable zeroValue(RelDataType type) {
     case SMALLINT:
     case INTEGER:
     case BIGINT:
+    case UTINYINT:
```
This function returns a comparable value, and not a literal. There is no promise that the type has literals.
I haven't checked whether this function is called elsewhere, perhaps the change is not necessary, but it is certainly harmless.

xiedeyantu commented on core/src/main/java/org/apache/calcite/rex/RexBuilder.java:
```diff
@@ -941,14 +942,30 @@ boolean canRemoveCastFromLiteral(RelDataType toType,
+    } else {
+      // This is e.g., an unsigned type
+      RelDataType i = typeFactory.createSqlType(SqlTypeName.INTEGER);
+      return makeAbstractCast(type, makeZeroValue(i));
```
The `makeAbstractCast` method with two entries has been marked as deprecated. You can use `makeAbstractCast(type, makeZeroValue(i), false)`.

mihaibudiu commented:
I have added a test for aggregation, which uncovered a bug in the validator, which I fixed (when building VALUES all the components should be cast to the same type; this only worked if all components were literals, but there are constants which are not literals).

xiedeyantu commented:
I have a small suggestion, I don’t know if it is appropriate or not. I understand that this PR will involve a lot of changes. Can it be listed as multiple tasks and implemented one by one? Otherwise, it may be difficult to merge in the short term. I'm not sure if it will impact existing functionality as we implement it one by one and merge it in.

mihaibudiu commented:
I don’t plan to add more commits in this one. I actually was hoping to merge it today. It should not affect any existing functionality, it only adds new stuff

xiedeyantu commented:
> I don’t plan to add more commits in this one. I actually was hoping to merge it today. It should not affect any existing functionality, it only adds new stuff

I agree with your idea. If there is no additional impact on existing functions, it is best to use the current commit as a temporary end point, otherwise this PR may have no end point.

mihaibudiu commented:
I will squash the commits in preparation for merging.
Please let me know if you think this needs more review.

xiedeyantu commented:
I think there may not need to do more review, two people have already approved it, and the subsequent commits are all to improve the functionality. If there are any imperfections, we can follow up with subtasks.
